### PR TITLE
Feature/apps 3025 initial collaboration view

### PR DESCRIPTION
**Pull request by bergieboy**
*WILL UPDATE BASE BRANCH WHEN APPS-2995 IS MERGED*

This adds the initial Collaborations View

Screen Shot 2021-07-08 at 6 53 16 PM

### CODE REVIEW

File: src/containers/collaborations/components/CreateCollaborationModal.js [hunk 1]
```diff
@@ -0,0 +1,161 @@
...
+            <Label htmlFor="new-collaboration-name">Name</Label>
+            <Input error={!isValidCollaborationName} onChange={handleOnChangeCollaborationName} />
+          </div>
+          <div>
+            <Label htmlFor="new-collaboration-description">Description</Label>
+            <Input onChange={handleOnChangeCollaborationDescription} />
+          </div>
+          <div>
+            <Label htmlFor="new-collaboration-description">Select organizations to add to collaboration</Label>
+            <Select
```
Comment (anzioka) [inline src/containers/collaborations/components/CreateCollaborationModal.js:128]: I believe the `htmlFor` value should be equal to the `id` attribute of the corresponding input element

File: src/containers/collaborations/components/CreateCollaborationModal.js [hunk 2]
```diff
@@ -0,0 +1,161 @@
...
+    }
+    else {
+      const orgIds = orgOptions.map((org) => org.value);
+      setCollaborationOrganizations(orgIds);
+    }
+  };
+
+  const handleOnClickPrimary = () => {
+    if (isNonEmptyString(collaborationTitle)) {
+      dispatch(
```
Comment (anzioka) [inline src/containers/collaborations/components/CreateCollaborationModal.js:74]: just checking, shouldn't the `collaborationName` be an nonEmptyString as well?

File: src/containers/collaborations/CollaborationsContainer.js [hunk 3]
```diff
@@ -0,0 +1,135 @@
...
+const CollaborationsContainer = () => {
+
+  const [isVisibleAddCollaborationModal, setIsVisibleCreateCollaborationModal] = useState(false);
+  const [paginationState, paginationDispatch] = useReducer(paginationReducer, INITIAL_PAGINATION_STATE);
+  const dispatch = useDispatch();
+  const collaborations :Map<UUID, Map> = useSelector(selectUsersCollaborations());
+  const filteredCollaborations = collaborations.filter((collaboration :Map, collaborationId :UUID) => (
+    collaboration && (paginationState.query === collaborationId
+      || get(collaboration, 'title', '').toLowerCase().includes(paginationState.query.toLowerCase()))
+  ));
```
[the remark this answers is not in the document]
Comment (UnsungHero97) [reply src/containers/collaborations/CollaborationsContainer.js:54]: I actually think matching against the id is super convenient, even if it's only taken advantage of internally. Sometimes I have an id and I just want to find the thing with that id. We already have this for matching org ids.

Using something like `.startsWith()` or `.includes()` over `===` might be better user experience.

File: src/containers/collaborations/components/CreateCollaborationModal.js [hunk 4]
```diff
@@ -0,0 +1,161 @@
...
+  };
+
+  const handleOnChangeCollaborationName = (event :SyntheticInputEvent<HTMLInputElement>) => {
+    setCollaborationName(event.target.value || '');
+    setIsValidCollaborationName(true);
+  };
+
+  const handleOnChangeCollaborationDescription = (event :SyntheticInputEvent<HTMLInputElement>) => {
+    setCollaborationDescription(event.target.value || '');
+  };
```
[the remark this answers is not in the document]
Comment (UnsungHero97) [reply src/containers/collaborations/components/CreateCollaborationModal.js:100]: Only thing I'd change is `useReducer` over `useState(fancyObject)`.

File: src/containers/collaborations/CollaborationsContainer.js [hunk 5]
```diff
@@ -0,0 +1,135 @@
...
+  const createNewCollaborationsRS :?RequestState = useRequestState([COLLABORATIONS, CREATE_NEW_COLLABORATION]);
+  const getCollaborationsRS :?RequestState = useRequestState([COLLABORATIONS, GET_COLLABORATIONS]);
+
+  useEffect(() => {
+    if (isStandby(getCollaborationsRS)) {
+      dispatch(getCollaborations());
+    }
+    if (isSuccess(createNewCollaborationsRS)) {
+      dispatch(getCollaborations());
+    }
```
[the remark this answers is not in the document]
Comment (anzioka) [reply src/containers/collaborations/CollaborationsContainer.js:78]: Also, I think it makes sense to call `dispatch(getCollaborations())` once and update the redux store after creating new collaborations.

File: src/containers/collaborations/CollaborationsContainer.js (hunk 3 above)
[the remark this answers is not in the document]
Comment (bergieboy) [reply src/containers/collaborations/CollaborationsContainer.js:54]: I'm going to keep the id filter and use `.includes()`

File: src/containers/collaborations/components/CreateCollaborationModal.js [hunk 6]
```diff
@@ -0,0 +1,221 @@
...
+      return state;
+  }
+};
+
+const CreateCollaborationModal = ({ onClose } :Props) => {
+
+  const dispatch = useDispatch();
+
+  const [isValidCollaborationTitle, setIsValidCollaborationTitle] = useState(true);
+  const [isValidCollaborationName, setIsValidCollaborationName] = useState(true);
```
Comment (anzioka) [inline src/containers/collaborations/components/CreateCollaborationModal.js:81]: the reducer state has `nameIsValid` and `titleIsValid`.

File: src/containers/collaborations/CollaborationsContainer.js [hunk 7]
```diff
@@ -0,0 +1,139 @@
...
+    paginationDispatch({ type: PAGE, page, start });
+  };
+
+  const createNewCollaborationsRS :?RequestState = useRequestState([COLLABORATIONS, CREATE_NEW_COLLABORATION]);
+  const getCollaborationsRS :?RequestState = useRequestState([COLLABORATIONS, GET_COLLABORATIONS]);
+
+  useEffect(() => {
+    if (isStandby(getCollaborationsRS) || isSuccess(createNewCollaborationsRS)) {
+      dispatch(getCollaborations());
+    }
```
Comment (anzioka) [inline src/containers/collaborations/CollaborationsContainer.js:75]: I still think  `dispatch(getCollaborations())` should only be called once, unless I'm missing something.
https://github.com/openlattice/lattice-orgs/pull/1150#discussion_r667135964

**Review by anzioka — DISMISSED** [review]

**Review by giuliacampana — DISMISSED** [review]

File: src/containers/collaborations/reducers/renameCollaborationDatabaseReducer.js [hunk 8]
```diff
@@ -0,0 +1,44 @@
...
+  return renameCollaborationDatabase.reducer(state, action, {
+    REQUEST: () => state
+      .setIn([RENAME_COLLABORATION_DATABASE, REQUEST_STATE], RequestStates.PENDING)
+      .setIn([RENAME_COLLABORATION_DATABASE, action.id], action),
+    SUCCESS: () => {
+      const storedSeqAction = state.getIn([RENAME_COLLABORATION_DATABASE, action.id]);
+      if (storedSeqAction) {
+        const { collaborationId, name } = storedSeqAction.value;
+        return state
+          .setIn([COLLABORATIONS, collaborationId, TITLE], name)
```
Comment (UnsungHero97) [inline src/containers/collaborations/reducers/renameCollaborationDatabaseReducer.js:29]: Renaming the database is independent of the title of the collaboration. You need to update the value in `DATABASE_DETAILS`.

**Review by UnsungHero97 — CHANGES_REQUESTED** [review]

**Review by UnsungHero97 — APPROVED** [review]

**Review by anzioka — APPROVED** [review]

**Review by giuliacampana — APPROVED** [review]

**Review by solomontang — APPROVED** [review]

**Pull request merged**